Add input and bounding shapes support (#2742)

xzfc opened the pull request:
Basic idea: if the window has a shape, set the parent container shape as the union of the window shape and the shape of the frame borders.

Related issue: #2742.

Thanks @psychon for the [initial work](https://github.com/i3/i3/issues/2742#issuecomment-433619265).

---

Screenshot of shaped `xeyes`, `oclock`, and `peek` floating windows:
![screenshot](http://dump.bitcheese.net/files/lorizij/2018-11-09_10:21:29.png)

---

1. New dependency: `xcb-shape`.
2. I check for Shape 1.1 support in `src/main.c`. Whereas it is possible to support Shape 1.0 (it does not have input shapes, only bounding shapes), but I think it does not worth that.
3. Shapes in tiled windows are weird: if shaped window is in tabbed/stacked container and `focus_follows_mouse` is enabled, it may be hard not to raise window underneath when moving mouse cursor.
4. Rendering of shaped tiled windows: the similar issue with windows titlebars as with transparent windows or in i3-gaps with titlebars.

Concerning 1, should I update `debian/control`?
Concerning 3 and 4, I'm not sure what i3 should do with shaped non-floating windows. Ignore shaping of tiled windows or leave it as is?

Airblader commented:
Thanks for the PR! This is really cool. I'll have a look at the PR a bit later, but regarding your questions:

1. Yeah, please update it. 
2. I think that's fine. Shape 1.1 predates i3 by years, so if you have i3 on your system, chances are your server supports no Shape or 1.1. And you only lose shape support otherwise, so yeah.
3. I think I'd accept that for now. I think chances are there's more usecases where it follows user intent than cases where it is a problem. 
4. That's more of a problem, I'll get back on this later as well as there's a few things I wanna check out first.

Airblader commented on src/handlers.c:
```diff
@@ -1517,6 +1518,21 @@ void handle_event(int type, xcb_generic_event_t *event) {
     }
 
+    if (shape_supported && type == shape_base + XCB_SHAPE_NOTIFY) {
+        DLOG("shape event.\n");
```
This log message is too unspecific. Please also log the affected window and the type of the event here.

Airblader commented on src/handlers.c:
```diff
@@ -1517,6 +1518,21 @@ void handle_event(int type, xcb_generic_event_t *event) {
+        DLOG("shape event.\n");
+        xcb_shape_notify_event_t *shape = (xcb_shape_notify_event_t *)event;
+        Con *con = con_by_window_id(shape->affected_window);
+        if (con == NULL)
```
I think we should log if this happens (again including the affected window). Also, please use `{} ` for all blocks.

Airblader commented on src/main.c:
```diff
@@ -683,6 +687,22 @@ int main(int argc, char *argv[]) {
+    } else {
+        shape_supported = false;
+    }
+    if (!shape_supported)
```
`{} `

Airblader commented on src/manage.c:
```diff
@@ -574,6 +574,11 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
      * cleanup) */
     xcb_change_save_set(conn, XCB_SET_MODE_INSERT, window);
 
+    /* Receive ShapeNotify events whenever the client altered it's window
```
nit: "its", not "it's"

Airblader commented on src/manage.c:
```diff
@@ -574,6 +574,11 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
+    /* Receive ShapeNotify events whenever the client altered it's window
+     * shape. */
+    if (shape_supported)
+        xcb_shape_select_input(conn, window, 1);
```
Please use `true` instead of `1` for booleans.

Airblader commented on src/manage.c:
```diff
@@ -574,6 +574,11 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
 
+    /* Receive ShapeNotify events whenever the client altered it's window
+     * shape. */
+    if (shape_supported)
```
`{}`

Airblader commented on src/tree.c:
```diff
@@ -248,6 +248,10 @@ bool tree_close_internal(Con *con, kill_window_t kill_window, bool dont_kill_par
 
+            /* Stop receiving ShapeNotify events. */
+            if (shape_supported)
+                xcb_shape_select_input(conn, con->window->id, 0);
```
`false` and `{}`

Airblader commented on src/x.c:
```diff
@@ -51,6 +51,9 @@ typedef struct con_state {
     bool need_reparent;
     xcb_window_t old_frame;
 
+    bool shaped;
```
Could we add a one line comment to each field which quickly describes what they represent?

Airblader commented on src/x.c:
```diff
@@ -730,6 +769,32 @@ static void set_hidden_state(Con *con) {
+    xcb_rectangle_t rectangles[4];
+    size_t rectangles_count = x_get_border_rectangles(con, rectangles);
+    if (rectangles_count)
+        xcb_shape_rectangles(conn, XCB_SHAPE_SO_UNION, shape_kind, XCB_CLIP_ORDERING_UNSORTED, con->frame.id, 0, 0, rectangles_count, rectangles);
```
`{}`

Airblader commented on src/x.c:
```diff
@@ -793,6 +860,16 @@ void x_push_node(Con *con) {
         DLOG("ignore_unmap for reparenting of con %p (win 0x%08x) is now %d\n",
              con, con->window->id, con->ignore_unmap);
+
+        need_reshape = true;
```
Why do we need to reshape here?

xzfc replied:
We are applying the shape on the parent window. When the parent is changed, we need to apply the shape on the new parent.

Airblader commented on src/x.c:
```diff
@@ -898,6 +975,14 @@ void x_push_node(Con *con) {
 
+    if (shape_supported && need_reshape && con->window != NULL) {
+        /* We need to reshape the window frame only if it already have shape. */
+        if (state->shaped)
```
`{}` for both statements

Airblader commented on src/x.c:
```diff
@@ -898,6 +975,14 @@ void x_push_node(Con *con) {
     }
 
+    if (shape_supported && need_reshape && con->window != NULL) {
+        /* We need to reshape the window frame only if it already have shape. */
```
nit: "[...] only if it already has a shape."

Airblader commented on src/x.c:
```diff
@@ -1325,3 +1410,26 @@ void x_mask_event_mask(uint32_t mask) {
+void x_set_shape(Con *con, xcb_shape_sk_t kind, bool enable) {
+    struct con_state *state;
+    if ((state = state_for_frame(con->frame.id)) == NULL) {
+        ELOG("window state for con not found\n");
```
Please include the ID of the con here

Airblader replied:
Yeah, but that doesn't mean in new code we can't do better. Having IDs in log messages makes reading the log files a whole lot easier. :-)

xzfc replied:
Maybe we should add ID logging into `state_for_frame` instead? It is already have `ELOG("No state found\n")` inside.

Airblader replied:
Yeah, that would also be good. If we log here as well, we should still add it here, too, though.

Airblader commented on src/x.c:
```diff
@@ -1325,3 +1410,26 @@ void x_mask_event_mask(uint32_t mask) {
+        return;
+    }
+
+    assert(kind == XCB_SHAPE_SK_INPUT || kind == XCB_SHAPE_SK_BOUNDING);
```
I think an assert is overly aggressive here. I'd just switch this construct so a switch case:

```
switch (kind) {
    case XCB_SHAPE_SK_INPUT:
        state->input_shaped = enable;
        break;
    case XCB_SHAPE_SK_BOUNDING:
        state->shaped = enable;
        break;
    default:
        // Use a log instead, also log con ID 
        ELOG("received unknown shape event type for con = ... This is a bug in i3\n");
        break;
}

if (enable) {
    x_shape_frame(con, kind);
} else {
    x_unshape_frame(con, kind);
} 
```

Airblader commented:
FYI, for question 4 #1966 would be the solution, but that isn't exactly a trivial change. I can't see a good way to prevent this issue with simple means, so I think for now we'd have to disable Shape for tiled windows.

xzfc commented:
Updated regarding to the comments, added missing `xcb_flush()` in `x_set_shape()`, and made shaping only for floating windows.

orestisf1993 commented on src/x.c:
```diff
@@ -898,6 +985,28 @@ void x_push_node(Con *con) {
+            }
+        }
+    }
+    state->was_floating = con->parent->type == CT_FLOATING_CON;
```
This also segfaults my setup upon starting with xephyr. Got to go, no time to debug but here's a backtrace:

```
#0 0x5570b5f512c1 in x_push_node ../../i3/src/x.c:1008
#1 0x5570b5f547d2 in x_push_changes ../../i3/src/x.c:1229
#2 0x5570b5f2730e in tree_render ../../i3/src/tree.c:463
#3 0x5570b5effc94 in randr_query_outputs ../../i3/src/randr.c:955
#4 0x5570b5f01b1d in randr_init ../../i3/src/randr.c:1091
#5 0x5570b5ed462f in main ../../i3/src/main.c:755
```

xzfc replied:
Fixed segfault.

xzfc commented on src/handlers.c:
```diff
@@ -1517,6 +1518,27 @@ void handle_event(int type, xcb_generic_event_t *event) {
+            return;
+        }
+
+        if (shape->shape_kind == XCB_SHAPE_SK_BOUNDING ||
```
We need to track that tiled container has shape and apply the shape later when it becomes floating.

Airblader replied:
Yeah, those changes wouldn't affect tracking that, would they?

xzfc replied:
`state->shape` is set inside `x_set_shape` before checking if window is floating.

Airblader replied:
Ah, right, I see now. My bad! Then in x_set_shape we should also use con_is_floating, though.

xzfc commented:
New update.

* Changes regarding to the new comments.
* Moved `shaped`, `input_shaped` booleans from `struct con_state` (`src/x.c`) to `struct Window` (`include/data.h`) as it reflects the state of application window, not the state of the container.
* Initial check for the shape in `manage_window` (to handle case when the window sets the shape once and newer updates it).
* Fixed segfault.

Airblader commented:
@xzfc Great, thanks! Could you also rebase to resolve the conflict?

orestisf1993 commented:
I can take a look but not immediately. Let's say an upper bound of one week?

xzfc commented:
Travis still fails with `E: Unable to locate package clang-format-4.0`. I think it's related to #3497.
https://travis-ci.org/i3/i3/jobs/456670878#L557

Airblader commented:
Hm, that's certainly odd because I just restarted a Travis build on next and it goes through fine. I can't investigate it at the moment, though.

Airblader commented:
@stapelberg Do you happen to know why clang-format-4.0 can't be found when building this PR? Searching for it it does seem like it's not in debian:sid, but then how does it work on the next branch?

stapelberg commented:
Does it help to rebase the PR onto next? If no, let me know and I’ll take a look this evening.

Airblader commented:
We rebased a few days ago and it failed then, too. I don't think we changed anything in the time since either.

stapelberg commented:
Turns out clang-format-4.0 vanished from Debian testing, too, see https://packages.debian.org/search?keywords=clang-format-4.0&searchon=names&suite=all&section=all

Essentially, we’ll need another commit such as https://github.com/i3/i3/commit/eedb25765590460d328f42364cc2396a6fcd9f0e#diff-7985139ae9b6efb45373e3e36e444224, but with clang-format-6.0, I suppose?

stapelberg commented:
Try rebasing to next now.

orestisf1993 commented on src/x.c:
```diff
@@ -396,6 +401,55 @@ static void x_draw_decoration_after_title(Con *con, struct deco_render_params *p
     x_draw_title_border(con, p);
 }
 
+/* Get rectangles representing the border around the child window. Some borders
```
```suggestion
/*
 * Get rectangles representing the border around the child window. Some borders
```

orestisf1993 commented on src/x.c:
```diff
@@ -396,6 +401,55 @@ static void x_draw_decoration_after_title(Con *con, struct deco_render_params *p
 
+/* Get rectangles representing the border around the child window. Some borders
+ * are adjacent to the screen-edge and thus not returned. Return value is the
+ * number of rectangles. */
```
```suggestion
 * number of rectangles.
 *
 */
```

xzfc commented:
* Added test case.
* Rebased onto next. Travis is now green. ✅
* Applied the suggested change by @orestisf1993.

algmyr commented:
Just wanted to leave a thank you for everyone who worked on the feature. It seems to work perfectly, ignoring non-floating windows as was mentioned earlier. Superb work!
